File: app/src/main/java/lib/mozidev/me/extextview/MainActivity.java
package lib.mozidev.me.extextview;

import android.graphics.Color;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.Gravity;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.Switch;

import java.util.Arrays;
import java.util.List;

import lib.mozidev.me.extendedtextview.ExtendedTextView;
import lib.mozidev.me.extendedtextview.StrikeThroughPainting;

public class MainActivity extends AppCompatActivity {

    private static final String SINGLE_LINE_TEXT = "I'm making a note here: Huge success!";
    private static final String MULTIPLE_LINES_TEXT = "With '\\n' wrap\n" +
            "Aperture Science:\n" +
            "We do what we must\n" +
            "because we can\n" +
            "For the good of all of us.\n" +
            "Except the ones who are dead.";
    private static final String VERY_LONG_TEXT = "Early work on the Aperture Science Handheld Portal Device began; the early version, called the Aperture Science Portable Quantum Tunneling Device, proved to be too bulky for effective use...";
    private static final List<String> strs = Arrays.asList(SINGLE_LINE_TEXT, MULTIPLE_LINES_TEXT, VERY_LONG_TEXT);
    private int strIndex = 0;

    private float strokeWidth = 2F;
    private int strokeColor = Color.BLACK;
    private boolean cutEdge = true;

    private StrikeThroughPainting strikeThroughPainting = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final ExtendedTextView tv = findViewById(R.id.extendedTextView);
        strikeThroughPainting = new StrikeThroughPainting(tv);

        findViewById(R.id.button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                strIndex += 1;
                if (strIndex == strs.size()) {
                    strIndex = 0;
                }
                tv.setText(strs.get(strIndex));
            }
        });

        findViewById(R.id.button2).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                strikeThroughPainting
                        .cutTextEdge(cutEdge)
                        .color(strokeColor)
                        .strokeWidth(strokeWidth)
                        .mode(StrikeThroughPainting.MODE_DEFAULT)
                        .strikeThrough();
            }
        });

        findViewById(R.id.button3).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                strikeThroughPainting
                        .cutTextEdge(cutEdge)
                        .color(strokeColor)
                        .strokeWidth(strokeWidth)
                        .mode(StrikeThroughPainting.MODE_LINES_TOGETHER)
                        .strikeThrough();
            }
        });

        Switch switch1 = findViewById(R.id.switch1);
        switch1.setChecked(cutEdge);
        switch1.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                cutEdge = isChecked;
            }
        });

        this.<Switch>findViewById(R.id.switch2).setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    tv.setLineSpacing(27F, 2F);
                } else {
                    tv.setLineSpacing(0, 1);
                }
            }
        });

        this.<Switch>findViewById(R.id.switch3).setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    strokeColor = Color.BLUE;
                    strokeWidth = 6F;
                } else {
                    strokeColor = Color.BLACK;
                    strokeWidth = 2F;
                }
            }
        });

        this.<Switch>findViewById(R.id.switch4).setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    tv.setGravity(Gravity.CENTER_HORIZONTAL);
                } else {
                    tv.setGravity(Gravity.NO_GRAVITY);
                }
            }
        });

        findViewById(R.id.button4).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                strikeThroughPainting.clearStrikeThrough();
            }
        });
    }
}

File: sample/src/main/java/lib/mozidev/me/demo/MainActivity.kt
package lib.mozidev.me.demo

import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar
import lib.mozidev.me.demo.databinding.ActivityMainBinding
import lib.mozidev.me.extextview.ExTextView
import lib.mozidev.me.extextview.StrikeThroughPainting

class MainActivity : AppCompatActivity() {
    private var strIndex = 0
    private var strokeWidth = 2f
    private var strokeColor = Color.BLACK
    private var cutEdge = true
    private lateinit var strikeThroughPainting: StrikeThroughPainting

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        reset(binding.extendedTextView)
        binding.button.setOnClickListener {
            reset(binding.extendedTextView)
            strIndex += 1
            if (strIndex == strs.size) {
                strIndex = 0
            }
            binding.extendedTextView.text = strs[strIndex]
        }
        binding.button2.setOnClickListener {
            reset(binding.extendedTextView)
            strikeThroughPainting
                .cutTextEdge(cutEdge)
                .color(strokeColor)
                .strokeWidth(strokeWidth)
                .mode(StrikeThroughPainting.MODE_DEFAULT)
                .strikeThrough()
        }
        binding.button3.setOnClickListener {
            reset(binding.extendedTextView)
            strikeThroughPainting
                .cutTextEdge(cutEdge)
                .color(strokeColor)
                .strokeWidth(strokeWidth)
                .mode(StrikeThroughPainting.MODE_LINES_TOGETHER)
                .strikeThrough()
        }
        binding.switch1.isChecked = cutEdge
        binding.switch1.setOnCheckedChangeListener { _, isChecked ->
            cutEdge = isChecked
            reset(binding.extendedTextView)
        }
        binding.switch2.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                binding.extendedTextView.setLineSpacing(27f, 2f)
            } else {
                binding.extendedTextView.setLineSpacing(0f, 1f)
            }
            reset(binding.extendedTextView)
        }
        binding.switch3.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                strokeColor = Color.BLUE
                strokeWidth = 6f
            } else {
                strokeColor = Color.BLACK
                strokeWidth = 2f
            }
            reset(binding.extendedTextView)
        }
        binding.switch4.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                binding.extendedTextView.gravity = Gravity.CENTER_HORIZONTAL
            } else {
                binding.extendedTextView.gravity = Gravity.NO_GRAVITY
            }
            reset(binding.extendedTextView)
        }
        binding.button5.setOnClickListener {
            reset(binding.extendedTextView)
            strikeThroughPainting
                .cutTextEdge(cutEdge)
                .color(strokeColor)
                .strokeWidth(strokeWidth)
                .linePosition(0.65f)
                .callback { Snackbar.make(
                    findViewById(R.id.container),
                    "Callback after animation", Snackbar.LENGTH_LONG
                ).show()}
                .strikeThrough()
        }
        binding.button4.setOnClickListener { reset(binding.extendedTextView) }
    }

    // if not, we get funcky results when switching configurations
    private fun reset(targetView: ExTextView) {
        if (::strikeThroughPainting.isInitialized) {
            strikeThroughPainting.clearStrikeThrough()
        }
        strikeThroughPainting = StrikeThroughPainting(targetView)
    }

    companion object {
        private const val SINGLE_LINE_TEXT = "I'm making a note here: Huge success!"
        private const val MULTIPLE_LINES_TEXT = "With '\\n' wrap\n" +
                "Aperture Science:\n" +
                "We do what we must\n" +
                "because we can\n" +
                "For the good of all of us.\n" +
                "Except the ones who are dead."
        private const val VERY_LONG_TEXT =
            "Early work on the Aperture Science Handheld Portal Device began; the early version, called the Aperture Science Portable Quantum Tunneling Device, proved to be too bulky for effective use..."
        private val strs = listOf(SINGLE_LINE_TEXT, MULTIPLE_LINES_TEXT, VERY_LONG_TEXT)
    }
}
